### PR TITLE
Testing project

### PR DESCRIPTION
**Pull request by tjhei**
- put the tests into a separate cmake project
- do not recompile aspect while running ctest
- always use make as the generator for the test project

### CODE REVIEW

**Comment by tjhei** [comment]
make: 17m (3m for generating reference output, mostly spent scanning for dependencies)
ninja: 25m (11m for rerunning all tests to generate reference output)

So we would be at 14 minutes if we don't care about the reference output or we figure out a way to speed it up.

**Comment by tjhei** [comment]
It would be helpful if somebody could test this branch locally to figure out if tests are handled correctly or if I broke anything in a typical workflow when making/fixing/running tests.

**Comment by gassmoeller** [comment]
I did some tests, and it seems to work. I tested:
- replacing all test outputs with the script
- adding a new test with correct output (passes)
- modifying test output to wrong output (fails)
- correcting that test output again (passes)

Anything else that might not work?

**Comment by tjhei** [comment]
> I did some tests, and it seems to work

Thanks. Did you use ninja or make?

> Anything else that might not work?

Not sure to be honest. I think the quicktest needs some testing.

**Comment by gassmoeller** [comment]
I used make for all testing, and cmake 3.6.1 + 2.8.12.

**Comment by tjhei** [comment]
Okay, I am now always using make as the generator for the test project because ninja is messing up parallel builds. I think I have a way to work around this on the tester though.

I think this is ready for review/merging.

File: CMakeLists.txt [hunk 1]
```diff
@@ -135,9 +135,30 @@ ADD_CUSTOM_TARGET(setup_tests
   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
   COMMENT "Enabling all tests ...")
 
+SET(ASPECT_TEST_GENERATOR "Unix Makefiles" CACHE STRING "Generator to use for the test cmake project. Using ninja instead of make is not recommended.")
+
 IF(EXISTS ${CMAKE_SOURCE_DIR}/tests/CMakeLists.txt)
-  ENABLE_TESTING()
-  ADD_SUBDIRECTORY(tests)
+  FILE(WRITE ${CMAKE_BINARY_DIR}/CTestTestfile.cmake "SUBDIRS(tests)")
```
Comment (bangerth) [inline CMakeLists.txt]: I think you ought to quote the filename

Comment (bangerth) [inline CMakeLists.txt]: although we seem to not do this anywhere else either

File: CMakeLists.txt [hunk 2]
```diff
@@ -135,9 +135,30 @@ ADD_CUSTOM_TARGET(setup_tests
...
+
+  FILE(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/tests)
+
+  SET(aspect_binary ${CMAKE_BINARY_DIR}/aspect)
+
+  MESSAGE(STATUS "Setting up test project, see setup_tests.log for details.")
+  EXECUTE_PROCESS(
+    COMMAND ${CMAKE_COMMAND} -G ${ASPECT_TEST_GENERATOR}
```
Comment (bangerth) [inline CMakeLists.txt]: in your example, `ASPECT_TEST_GENERATOR` is a string with spaces. What happens when it is passed like this? Will the command line take this apart into separate strings?

File: CMakeLists.txt [hunk 3]
```diff
@@ -135,9 +135,30 @@ ADD_CUSTOM_TARGET(setup_tests
   WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
   COMMENT "Enabling all tests ...")
 
+SET(ASPECT_TEST_GENERATOR "Unix Makefiles" CACHE STRING "Generator to use for the test cmake project. Using ninja instead of make is not recommended.")
```
Comment (bangerth) [inline CMakeLists.txt]: maybe break the doc string onto the next line

**Comment by bangerth** [comment]
I can't say that I really understand what you are doing here. Am I correct that the problem only happens when the ASPECT executable is dirty (needs to be re-built) and we run ctest? That then every test tries to rebuilt the ASPECT executable, which may per chance work with `make` but not with `ninja`? And that you address this by (1) making the testsuite a separate project that doesn't know how to rebuilt the executable, and (2) using make to avoid the race conditions?

I can understand (1). Conceptually, I think it would actually be even nicer if we could do the following: instead of listing the ASPECT executable as a dependency, test that the executable is up-to-date and otherwise fail. I don't know how to test that in cmake, though. It would be a nice check if there was a way to find out whether a target is clean or dirty, without attempting to re-build it. 

But if you go with (1), then the testsuite will never try to rebuild the executable (because that is built by a separate project, and the testsuite no longer knows how to rebuild it). So we may be using an old executable, which would be unfortunate. But furthermore, why do we then need to insist on using `make` over `ninja`?

**Comment by tjhei** [comment]
> Am I correct that the problem only happens when the ASPECT executable is dirty (needs to be re-built) and we run ctest?

Not quite. There are several things happening:
1. To rerun each test if needed (regenerate the screen-output), we need this target to depend on the aspect binary. If they are both in the same cmake project (like it is without this PR), cmake will automatically add a dependency to the cmake target ``aspect`` (instead of just checking the timestamp of the binary). This has several consequences:
    - If you use make it will take something like 10s to check if ``aspect`` is out of date (scanning all .h and .cc file stamps etc.). This happens for every single test.
    - If you use ``ctest -j X`` and aspect is out of date, you have make running X times in parallel deciding at the same time to rebuild. This will obviously fail (and crash your computer if you use ninja because it will spawn X*NPROC compile tasks).
2. If you run ``ctest -j X``, ctest will spawn X processes to make tests at the same time. Ninja does not support running in the same project concurrently ( https://groups.google.com/d/msg/ninja-build/4VP7whvWSH8/gXbMTWfaxp0J ). You get warnings ``ninja: warning: premature end of file; recovering`` and it will not track dependencies correctly (rebuilding stuff over and over for example). In short: we can not use ninja for targets that might be executed with ``ctest -j X``. That is why we made the switch in deal.II.

By breaking the two projects and always using ``make`` for the test project even if you are using ninja to compile your source code, we can fix both things.

> I think it would actually be even nicer if we could do the following: instead of listing the ASPECT executable as a dependency, test that the executable is up-to-date and otherwise fail.

This won't work because checking the dependencies over and over is too slow.

> But if you go with (1), then the testsuite will never try to rebuild the executable (because that is built by a separate project, and the testsuite no longer knows how to rebuild it). So we may be using an old executable, which would be unfortunate.

Correct. I don't see a way around that, unfortunately.

**Comment by tjhei** [comment]
Okay, I figured out a clever workaround for the tester. We are down to 12 minutes total with this PR.

File: CMakeLists.txt (hunk 1 above)
Comment (tjhei) [inline CMakeLists.txt]: cmake quotes arguments automatically (tried with a build directory containing a space)

File: CMakeLists.txt (hunk 2 above)
Comment (tjhei) [inline CMakeLists.txt]: Works because EXECUTE_PROCESS will escape arguments automatically (VERBATIM option).

**Comment by bangerth** [comment]
OK, thanks for the explanation. Good to merge if you think it's ready (and if we can ignore the 8.5 tester).

**Comment by tjhei** [comment]
> Good to merge if you think it's ready (and if we can ignore the 8.5 tester).

#1463 broke mainline...

**Comment by bangerth** [comment]
On 04/29/2017 03:44 PM, Timo Heister wrote:
> #1463 <https://github.com/geodynamics/aspect/pull/1463> broke mainline...

Well, but now...

**Comment by tjhei** [comment]
rebased.

**Comment by bangerth** [comment]
OK. I can't quite figure out whether you think that it is ready and is merged, or not. I'm ok with merging. If you think it's ready, hit the button.

**Pull request merged**